src/istio/mixerclient/client_impl.cc

```diff
@@ -52,110 +52,122 @@ MixerClientImpl::MixerClientImpl(const MixerClientOptions &options)
+      // implementation will still need to send a request to the Mixer server
+      // in the background.
+      //
+      context->setFinalStatus(context->quotaStatus());
+      on_done(*context);
+      on_done = nullptr;
+      if (!context->remoteQuotaRequestRequired()) {
+        return nullptr;
+      }
     }
   }
 
-  compressor_.Compress(attributes, request->mutable_attributes());
-  request->set_global_word_count(compressor_.global_word_count());
-  request->set_deduplication_id(deduplication_id_base_ +
-                                std::to_string(deduplication_id_.fetch_add(1)));
-
-  // Need to make a copy for processing the response for check cache.
-  Attributes *attributes_copy =
-      google::protobuf::Arena::CreateMessage<Attributes>(arena);
-  CheckResponse *response =
-      google::protobuf::Arena::CreateMessage<CheckResponse>(arena);
-  *attributes_copy = attributes;
-  // Lambda capture could not pass unique_ptr, use raw pointer.
-  CheckCache::CheckResult *raw_check_result = check_result.release();
-  QuotaCache::CheckResult *raw_quota_result = quota_result.release();
+  // TODO(jblatt) mjog thinks this is a big CPU hog.  Look into it.
+  context->compressRequest(
+      compressor_,
+      deduplication_id_base_ + std::to_string(deduplication_id_.fetch_add(1)));
+
   if (!transport) {
     transport = options_.env.check_transport;
   }
+
+  //
   // We are going to make a remote call now.
+  //
+
   ++total_remote_check_calls_;
-  if (!quotas.empty()) {
+
+  if (context->quotaCheckRequired()) {
     ++total_remote_quota_calls_;
   }
+
   if (on_done) {
     ++total_blocking_remote_check_calls_;
-    if (!quotas.empty()) {
+    if (context->quotaCheckRequired()) {
       ++total_blocking_remote_quota_calls_;
     }
   }
 
-  return transport(
-      *request, response,
-      [this, attributes_copy, response, raw_check_result, raw_quota_result,
-       on_done, arena](const Status &status) {
-        raw_check_result->SetResponse(status, *attributes_copy, *response);
-        raw_quota_result->SetResponse(status, *attributes_copy, *response);
-        CheckResponseInfo check_response_info;
-        if (on_done) {
-          if (!raw_check_result->status().ok()) {
-            check_response_info.response_status = raw_check_result->status();
-          } else {
-            check_response_info.response_status = raw_quota_result->status();
-          }
-          check_response_info.route_directive =
-              raw_check_result->route_directive();
-          on_done(check_response_info);
-        }
-        delete raw_check_result;
-        delete raw_quota_result;
-        delete arena;
-
-        if (utils::InvalidDictionaryStatus(status)) {
-          compressor_.ShrinkGlobalDictionary();
-        }
-      });
+  return transport(context->request(), context->response(),
+                   [this, context = context, on_done](const Status &status) {
```

Comment:
is this lambda invoked immediately in transport, or scheduled later? If latter, `*this` may be destructed.

Reply:
Lamda may be invoked later.   The this pointer refers to the MixerClientImpl object which has a much longer lifetime than any individual transaction (in this case the CheckContext has transaction lifespan).    

But fundamentally I'm relying on no transactions completing after the MixerClientImpl is destroyed like the previous code and it's possible there may still be issues there.